common/event_emitter_test.go

```diff
@@ -125,3 +127,125 @@ func TestEventEmitterAllEvents(t *testing.T) {
 		})
 	})
 }
+
+func TestBaseEventEmitter(t *testing.T) {
+	t.Parallel()
+
+	t.Run("order of emitted events kept", func(t *testing.T) {
+		// Test description
+		//
+		// 1. Emit many events from the emitWorker.
+		// 2. Handler receives the emitted events.
+		//
+		// Success criteria: Ensure that the ordering of events is
+		//                   received in the order they're emitted.
+
+		t.Parallel()
+
+		eventName := "AtomicIntEvent"
+		maxInt := 100
+
+		ctx, cancel := context.WithCancel(context.Background())
+		emitter := NewBaseEventEmitter(ctx)
+		ch := make(chan Event)
+		emitter.on(ctx, []string{eventName}, ch)
+
+		var expectedI int
+		handler := func() {
+			defer cancel()
+
+			for expectedI != maxInt {
+				e := <-ch
+
+				i, ok := e.data.(int)
+				if !ok {
+					assert.FailNow(t, "unexpected type read from channel", e.data)
+				}
+
+				assert.Equal(t, eventName, e.typ)
+				assert.Equal(t, expectedI, i)
+
+				expectedI++
+			}
+
+			close(ch)
+		}
+		go handler()
+
+		emitWorker := func() {
+			for i := 0; i < maxInt; i++ {
+				i := i
+				emitter.emit(eventName, i)
+			}
+		}
+		go emitWorker()
+
+		select {
+		case <-ctx.Done():
+		case <-time.After(time.Second * 2):
+			assert.FailNow(t, "test timed out, deadlock?")
+		}
+	})
+
+	t.Run("handler can emit without deadlocking", func(t *testing.T) {
+		// Test description
+		//
+		// 1. Emit many events from the emitWorker.
+		// 2. Handler receives emitted events (AtomicIntEvent1).
+		// 3. Handler emits event as AtomicIntEvent2.
+		// 4. Handler received emitted events again (AtomicIntEvent2).
+		//
+		// Success criteria: No deadlock should occur between receiving,
+		//                   emitting, and receiving of events.
+
+		t.Parallel()
+
+		eventName1 := "AtomicIntEvent1"
+		eventName2 := "AtomicIntEvent2"
+		maxInt := 100
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
```

Comment:
The timeout was based on your earlier suggestion:

> Nit: this test doesn't assert much, but if we're not expecting a deadlock then it should fail after a specific timeout. Otherwise we run the risk of timing out whenever the global test timeout is reached.

I prefer to remove it and revert back to `WithCancel()`. If there's a deadlock it should show up when the global test timeout fires.

Reply:
Done 👍